flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushFilterIntoTableSourceScanAcrossWatermarkRule.java

```diff
@@ -0,0 +1,153 @@
+                FlinkRelFactories.FLINK_REL_BUILDER(),
+                "PushFilterIntoTableSourceScanAcrossWatermarkRule");
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        LogicalFilter filter = call.rel(0);
+        LogicalWatermarkAssigner watermarkAssigner = call.rel(1);
+        LogicalTableScan scan = call.rel(2);
+
+        TableSourceTable oldTableSourceTable = scan.getTable().unwrap(TableSourceTable.class);
+
+        RelBuilder relBuilder = call.builder();
+        FlinkContext context = ShortcutUtils.unwrapContext(scan);
+        int maxCnfNodeCount = FlinkRelOptUtil.getMaxCnfNodeCount(scan);
+        RexNodeToExpressionConverter converter =
+                new RexNodeToExpressionConverter(
+                        relBuilder.getRexBuilder(),
+                        filter.getInput().getRowType().getFieldNames().toArray(new String[0]),
+                        context.getFunctionCatalog(),
+                        context.getCatalogManager(),
+                        TimeZone.getTimeZone(context.getTableConfig().getLocalTimeZone()));
+
+        Tuple2<RexNode[], RexNode[]> tuple2 =
+                RexNodeExtractor.extractConjunctiveConditions(
+                        filter.getCondition(),
+                        maxCnfNodeCount,
+                        relBuilder.getRexBuilder(),
+                        converter);
+        RexNode[] convertiblePredicates = tuple2._1;
+        RexNode[] unconvertedPredicates = tuple2._2;
+        if (convertiblePredicates.length == 0) {
+            // no condition can be translated to expression
+            return;
+        }
+
+        // record size before applyFilters for update statistics
+        int originPredicatesSize = convertiblePredicates.length;
+
+        // update DynamicTableSource
+        DynamicTableSource newTableSource = oldTableSourceTable.tableSource().copy();
+
+        SupportsFilterPushDown.Result result =
+                FilterPushDownSpec.apply(
+                        Arrays.asList(convertiblePredicates),
+                        newTableSource,
+                        SourceAbilityContext.from(scan));
+
+        relBuilder.push(scan);
+        List<RexNode> acceptedPredicates =
+                convertExpressionToRexNode(result.getAcceptedFilters(), relBuilder);
+        FilterPushDownSpec filterPushDownSpec = new FilterPushDownSpec(acceptedPredicates);
+
+        // record size after applyFilters for update statistics
+        int updatedPredicatesSize = result.getRemainingFilters().size();
+        // set the newStatistic newTableSource and extraDigests
+        TableSourceTable newTableSourceTable =
+                oldTableSourceTable.copy(
+                        newTableSource,
+                        getNewFlinkStatistic(
+                                oldTableSourceTable, originPredicatesSize, updatedPredicatesSize),
+                        getNewExtraDigests(result.getAcceptedFilters()),
```

Comment:
I tried pulling it into a reusable base class, but since it's all side effects being performed here it wasn't very intuitive. I'll give it another try.